File: brainfuck.clj
(defn bfout [c] (printf (str (char c))))
(defn bfin [] (flush) (byte (nth (read-line) 0)))

(defn retrieve-left [code ind amount]
    (if (= amount 0)
        (+ ind 1)
        (let [c (nth code (- ind 1))]
            (if (= c \[)
                (recur code (- ind 1)(- amount 1))
                (if (= c \])
                    (recur code (- ind 1)(+ amount 1))
                    (recur code (- ind 1) amount))))))

(defn retrieve-right [code ind amount]
    (if (= amount 0)
        ind
        (let [c (nth code (+ ind 1))]
            (if (= c \])
                (recur code (+ ind 1)(- amount 1))
                (if (= c \[)
                    (recur code (+ ind 1)(+ amount 1))
                    (recur code (+ ind 1) amount))))))

(defn bf-run [code iptr data dptr len]
  (when (< iptr len)
    (let [c (nth code iptr)]
      (cond
        (= c \>)
            (recur code (+ iptr 1) data (+ dptr 1) len)
        (= c \<)
            (recur code (+ iptr 1) data (- dptr 1) len)
        (= c \+)
            (recur code (+ iptr 1) (assoc data dptr (+ (nth data dptr nil) 1)) dptr len)
        (= c \-)
            (recur code (+ iptr 1) (assoc data dptr (- (nth data dptr nil) 1)) dptr len)
        (= c \.)
            (do
              (bfout (nth data dptr nil))
              (recur code (+ iptr 1) data dptr len))
        (= c \,)(recur code (+ iptr 1) (assoc data dptr (bfin)) dptr len)
        (= c \[)
            (if (= (nth data dptr) 0)
                (recur code (retrieve-right code iptr 1) data dptr len)
                (recur code (+ iptr 1) data dptr len))
        (= c \])
            (if (not= (nth data dptr) 0)
                (recur code (retrieve-left code (- iptr 1) 1) data dptr len)
                (recur code (+ iptr 1) data dptr len))
        :else (recur code (+ iptr 1) data dptr len)))))
  
(def code (slurp "tests/helloworld.bf"))
(bf-run code 0 (vec (make-array Integer/TYPE 30000)) 0 (count code))

File: brainfuck.java
import java.io.IOException;
import java.nio.file.*;

class Main {
    
    public static class BrainFuck {
        private String code;
        private int iptr;
        private byte[] data;
        private int dptr;

        public BrainFuck(String c){
            this.code = c;
            this.iptr = 0;
            this.data = new byte[30000];
            this.dptr = 0;
        }

        public void run() throws IOException {
            while(this.iptr < this.code.length()){
                int count = 1;
                switch(this.code.charAt(this.iptr)){
                    case '>':
                        this.dptr++;
                        break;
                    case '<':
                        this.dptr--;
                        break;
                    case '+':
                        this.data[this.dptr]++;
                        break;
                    case '-':
                        this.data[this.dptr]--;
                        break;
                    case '.':
                        System.out.print(Character.toString((char)this.data[this.dptr]));
                        break;
                    case ',':
                        this.data[this.dptr] = (byte)System.in.read();
                        break;
                    case '[':
                        count = 1;
                        if(this.data[this.dptr] == 0){
                            this.iptr++;
                            while(count > 0){
                                if(this.code.charAt(this.iptr) == '['){
                                    count++;
                                }
                                else if(this.code.charAt(this.iptr) == ']'){
                                    count--;
                                }
                                this.iptr++;
                            }
                            this.iptr--;
                        }
                        break;
                    case ']':
                        count = 1;
                        if(this.data[this.dptr] != 0){
                            this.iptr--;
                            while(count > 0){
                                if(this.code.charAt(this.iptr) == ']'){
                                    count++;
                                }
                                else if(this.code.charAt(this.iptr) == '['){
                                    count--;
                                }
                                this.iptr--;
                            }
                        }
                        break;
                    default:
                        break;
                }
                this.iptr++;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        String _c = new String(Files.readAllBytes(Paths.get(args[1])));
        BrainFuck bf = new Main.BrainFuck(_c);
        bf.run();
    }
}